Copy image to volume patch

saeki-masaki opened the pull request:
_is_cloneable / _parse_location の修正版パッチ
clone_image / local_path のコード修正追記

Commit:
環境習熟のテスト

Commit:
環境習熟のテスト

tishizaki committed:
sheepdog: fix to obtain the cloned volume from glance-image

Previously, invalid parse function of image_location was implemented.
Then, this patch changes followings
 - add common parse function of image_location
 - fix image_location format to 'sheep://<image_name>'
   as same as glance

Signed-off-by: Teruaki Ishizaki <ishizaki.teruaki@lab.ntt.co.jp>

tishizaki committed:
sheepdog: fix to obtain the cloned volume from glance-image

Previously, invalid parse function of image_location was implemented.
Then, this patch changes followings
 - add common parse function of image_location
 - fix image_location format to 'sheep://<image_name>'
   as same as glance

Signed-off-by: Teruaki Ishizaki <ishizaki.teruaki@lab.ntt.co.jp>

tishizaki committed:
sheepdog: fix bugs and coding style problem

Signed-off-by: Teruaki Ishizaki <ishizaki.teruaki@lab.ntt.co.jp>

tishizaki committed:
sheepdog: fix bugs and coding style problem

Signed-off-by: Teruaki Ishizaki <ishizaki.teruaki@lab.ntt.co.jp>

Commit:
Merge remote-tracking branch 'tishizaki/clone_from_image_to_volume' into sheepdog-devel

Commit:
Revert "環境習熟のテスト"

This reverts commit e02065c53396ce866088e7bc915b471164d7b830.

Commit:
remove non-ascii code contained in comments.

Commit:
Resolve confict and merge "_is_cloneable" and "_parse_location" to copy_image_to_volume branch

y-naga committed:
Add _run_qemu_img() in SheepdogCliet

y-naga committed:
Add tests for _run_qemu_img()

Commit:
modify _is_cloneable
  delete "clonable" and return directly True/False
  add log message when image_location is not startwith sheepdog:// etc.

Commit:
modify test_is_clonenable
   check LOG.info is called
   adjust location format as Glance
   add test case for _parse_location failed

Commit:
Merge remote-tracking branch 'sheepdog/sheepdog-devel' into local_newest

Conflicts:
	cinder/tests/unit/test_sheepdog.py
	cinder/volume/drivers/sheepdog.py

Commit:
Merge branch 'sheepdog-devel' of https://github.com/saeki-masaki/cinder into sheepdog-devel

Conflicts:
	cinder/volume/drivers/sheepdog.py

Commit:
change style probrems

Commit:
change using method defined in SheepdogDriverClass to SheepdogClientClass

Commit:
remove method _parse_location, it was move to SheepdogClientClass

Commit:
modify clone_image and create test code for.

Commit:
modify local_path and add test case

Commit:
modify local_path and add test case

Commit:
Merge branch 'copy_image_to_volume_3' into copy_image_to_volume_patch

Conflicts:
	cinder/volume/drivers/sheepdog.py

Commit:
Merge branch 'sheepdog-devel' of https://github.com/sheepdog/cinder into copy_image_to_volume_patch

Commit:
remove "dog" from command line , because use _run_dog interface

Commit:
add delete fail case in clone_volume
  add test case for above

tishizaki commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -36,6 +37,8 @@
 from cinder.volume import driver
 
+# set default snapshot name
+DEFAULT_SNAPNAME = 'snap'
```
サンプルコードで書いてた私の方でこう書いてしまってたんですが、DEFAULT_SNAPNAMEではなく、glanceが使うsnapshot名と分かるようにした方が良いかと思います。
また、sheepdog上からもglance-imageと分かるようなsnapshot名として, 例えばglance-imageとしてはいかがでしょうか？

saeki-masaki replied:
確かに視認性も考慮するとその方がよさそうですので、コメント反映します。
snap　-> glance-image

Commit:
change default snapshot name

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -115,6 +151,53 @@ def check_cluster_status(self):
+            raise exception.ImageUnacceptable(image_id=location, reason=reason)
+        if len(pieces) != 1:
+            reason = _('Not an sheepdog image')
+            raise exception.ImageUnacceptable(image_id=location, reason=reason)
```
パースとチェック処理についてはlambdaなど使わずにもう少し単純な方がよいかと思います。以下サンプル。

```
if not location.startswith(prefix):
  # プレフィックスが異なるエラー
pieces = location.lstrip(prefix).split('/')
if len(pieces) != 1:
  # 要素数が1以外のエラー
if len(pieces[0]) == 0:
  # VDI名が空の場合のエラー
```

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -115,6 +151,53 @@ def check_cluster_status(self):
+            reason = _('Not an sheepdog image')
+            raise exception.ImageUnacceptable(image_id=location, reason=reason)
+
+        return pieces
```
locationからVDI名を取得するメソッドとなっていますので、`return pieces[0]`で良いと思います。

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -115,6 +151,53 @@ def check_cluster_status(self):
+        # check whether volume is stored in sheepdog
+        (stdout, stderr) = self._run_dog('vdi', 'list', '-r', name)
+        if stdout is '':
+            LOG.debug('Image %s is not stored in Sheepdog', name)
```
細かいとこですが、ログメッセージで"in sheepdog"と"in Sheepdog"の大小文字が混在しています。

saeki-masaki replied:
sheepdog / Sheepdog の混在はログに限らずちらほらあるようです。

sheepプロセスやlocationのURLは小文字が原則なので、
案として Sheepdog を sheepdog に一括置換したいところ。

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -164,54 +247,33 @@ def __init__(self, *args, **kwargs):
+
+        try:
+            self.create_cloned_volume(volume, volume_ref)
+        except exception.VolumeBackendAPIException:
```
create_cloned_volume()を改修後にVolumeBackendAPIExceptionはSheepdogErrorの例外に置き換える。
#現段階ではTODO.

saeki-masaki replied:
TODOコメントを残した状態でコミットしてあります。（create_cloned_volumeの取り込みまで保留)

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -164,54 +247,33 @@ def __init__(self, *args, **kwargs):
+
+        try:
+            self._resize(volume, volume['size'])
+        except exception.VolumeBackendAPIException:
```
_resize()からはVolumeBackendAPIExceptionは投げられないため、例外が拾えません。

まずは、リサイズはself.client.resize()を利用するようにしてください。（まだsheepdog-develには含まれてませんが。）
その上で、例外についてはexcept Exceptionで全ての例外を拾ってロールバックを行った方が確実かと思います。
driverを利用するmanagerでもexcept Exceptionで拾うようになってました。

saeki-masaki replied:
Exceptionで例外を拾うように直しました。

y-naga commented on cinder/tests/unit/test_sheepdog.py:
```diff
@@ -262,6 +341,85 @@ def test_check_cluster_status_error_unknown(self):
+
+        with mock.patch.object(self.client, '_run_dog') as fake_execute:
+            with mock.patch.object(self.client,
+                                   '_parse_location') as parse_location:
```
mock.patch.objectのas変数名はfake_xxxで統一した方がいいかと思います。

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -306,7 +362,12 @@ def delete_snapshot(self, snapshot):
 
     def local_path(self, volume):
-        return "sheepdog:%s" % volume['name']
+        if volume['name'] is '':
```
中身の空の評価は== ''が良いのかもしれません。
_parse_location()では空チェックを==でやっていてこちらではisで行ってましたので。
文字列はどちらでもいいみたいですが。
参考：http://dackdive.hateblo.jp/entry/2014/08/07/172135

saeki-masaki replied:
直しました。

類似水平チェックが必要かと思われるコメントですので、各自再確認要ですね。

y-naga commented on cinder/volume/drivers/sheepdog.py:
```diff
@@ -306,7 +362,12 @@ def delete_snapshot(self, snapshot):
+        if volume['name'] is '':
+            msg = _('blank volume is not allowed')
+            LOG.error(msg)
+            raise exception.VolumeBackendAPIException(msg)
```
SheepdogDriverの中では、VolumeBackendAPIExceptionではなくSheepdogErrorを使う整理だったかと思います。
ただ、こちらに関してはSheepdogが絡んだエラーではないためVolumeBackendAPIException？？どうでしょう。。。

Commit:
variable 'name' change to 'volume_name'
  simplify location info check seqence in _parse_location
  use lower case for logging 'sheepdog'
  use statement " a == '' " for check blank string

  change unittest involved above.

saeki-masaki commented:
長原さんのコメントは一通り吸収済み。（1件他APIとの依存関係で保留)

Commit:
add condition checker when VDI is not snapshot
